### PR TITLE
Platform cleanup, std and pure-prove

### PR DESCRIPTION
**Pull request by shkoo**
* Fix rust prover sha bug, sha tests now pass

* Trim dependencies, less stuff to build in guest

* Add some Result propagation

* Guest methods can now decide whether they want to be no_std or not, and get the proper panic and alloc error implementations

* Merge separate "verify" method crate into regular "inner" method crate

### CODE REVIEW

File: risc0/zkvm/sdk/rust/guest/src/lib.rs [hunk 1]
```diff
@@ -35,34 +34,63 @@ pub mod sha;
...
+}
+
+/// This defines standalone (no_std) panic and alloc error handlers in
+/// a guest method, for guest methods that do not want to use the
+/// standard library.
+///
+/// Guest methods will also need to specify these at the beginning of the file:
+///
+/// #![no_std]
+/// #![feature(alloc_error_handler)]
```
Comment (flaub) [inline risc0/zkvm/sdk/rust/guest/src/lib.rs:69]: Won't this require guests to always use a nightly compiler even after we get the new target triple added to upstream?

File: risc0/zkvm/sdk/rust/src/prove/exec.rs [hunk 2]
```diff
@@ -751,7 +752,7 @@ where
                 self.code[CodeIndex::Data1High] = round.1;
             } else {
                 self.code[CodeIndex::ShaCtrl] = ONE;
-                self.code[CodeIndex::P1] = Fp::new((sha_phase - 64 + 4) as u32);
+                self.code[CodeIndex::P1] = Fp::new((sha_phase - 68 + 4) as u32);
```
Comment (flaub) [inline risc0/zkvm/sdk/rust/src/prove/exec.rs:755]: Nice find!

File: risc0/zkvm/sdk/rust/src/prove/mod.rs [hunk 3]
```diff
@@ -120,34 +120,35 @@ impl<'a> ProverImpl<'a> {
 }
 
 impl<'a> exec::IoHandler for ProverImpl<'a> {
-    fn on_txrx(&mut self, channel: u32, buf: &[u8]) -> Vec<u8> {
+    fn on_txrx(&mut self, channel: u32, buf: &[u8]) -> Result<Vec<u8>> {
```
Comment (flaub) [inline risc0/zkvm/sdk/rust/src/prove/mod.rs:123]: 👍

**Review by flaub — APPROVED** [review]

File: risc0/zkvm/sdk/rust/guest/src/lib.rs (hunk 1 above)
Comment (shkoo) [reply risc0/zkvm/sdk/rust/guest/src/lib.rs:69]: Only for no_std though

**Pull request merged**